[NFC] Rework types and terms for imports' access paths

beccadax opened the pull request:
Currently, the use of the term "access path" in relation to imports is inconsistent. It sometimes refers to the entire chain of dotted identifiers, sometimes just to the part scoping an import to a declaration, and occasionally refers to the part that *doesn't* scope an import to a declaration. This is made worse by the use of `typedef` and `using` in many places, making semantics and invariants unclear.

This PR reforms the terminology surrounding the dotted chain of identifiers used in `import` statements, and introduces new, strong, better-documented types to avoid mistakes or confusion about them. Specifically:

* An import path (the `ImportPath` type) represents an undifferentiated chain of identifiers in an `import` statement. By providing additional information about the `ImportKind`, it can be split into its constituent parts.
* A module path (the `ImportPath::Module` type) represents the part of an import path that identifies the module or submodule being imported.
* An access path (the `ImportPath::Access` type) represents the part of the import path that identifies the declaration targeted by a scoped import.

Each of these types has appropriate invariants about its length enforced, and a corresponding `::Builder` type can be used to create a temporary. I have also added a constructor to split apart strings into `Builder` types, and refactored away some code in SourceKit that used to duplicate the `ASTContext` method for handling dotted module names.

beccadax committed:
[NFC] Move ImportKind to a new Import.h header

beccadax committed:
[NFC] Add new ImportPath types

beccadax committed:
[NFC] Define “access path” in the lexicon

brentdax commented:
**[This is not yet ready to review.]**

**Review notes**: This is one of those patches that changes a ton of code, but most of the change is fairly mechanical. My suggestion is to review commit-by-commit. The mostly-mechanical adoption changes are in the third commit; you can probably skim through it, paying special attention to areas I have **[not yet]** commented.

beccadax committed:
[NFC] Get rid of ASTContext::getModuleByName dups

This method was reimplemented three times in various source tools.

beccadax committed:
[NFC] Add helper to parse import paths in strings

beccadax committed:
[NFC] Add ASTContext::getModuleByIdentifier()

This is a reasonably common operation.

beccadax committed:
[NFC] Rename “DeclPath” -> “AccessPath”

To avoid ambiguity, ImportResolution and a few other things used the term “decl path” instead of “access path”. Switch back to the correct terminology now that the compiler is becoming more consistent about it.

beccadax committed:
[NFC] Further simplify ModuleFile::associateWithFileContext

beccadax committed:
[NFC] Delete obsolete AccessPathTy-based code

This was all deprecated in earlier commits.

brentdax commented on lib/IDE/CodeCompletion.cpp:
```diff
@@ -1630,11 +1630,13 @@ class CodeCompletionCallbacksImpl : public CodeCompletionCallbacks {
+      // FIXME: Why are we allowing multiple module names? We don't support
+      // submodule qualification.
+      ImportPath::Module::Builder builder;
       for (auto Component : ITR->getComponentRange())
-        AccessPath.push_back({ Component->getNameRef().getBaseIdentifier(),
-                               Component->getLoc() });
-      if (auto Module = Context.getLoadedModule(AccessPath))
+        builder.push_back(Component->getNameRef().getBaseIdentifier(),
+                          Component->getLoc());
+      if (auto Module = Context.getLoadedModule(builder.get()))
```
I found the original code here confusing, so it's possible that I've misunderstood it somehow.

nathawes replied:
@brentdax I think you're right. It should probably just early exit when there's more than one component here.

brentdax replied:
I've restructured this to skip this logic unless there is exactly one component in the IdentTypeRepr.

brentdax commented on lib/IDE/ModuleInterfacePrinting.cpp:
```diff
@@ -374,8 +374,10 @@ static bool printModuleInterfaceDecl(Decl *D,
-  auto LHSPath = LHS->getFullAccessPath();
-  auto RHSPath = RHS->getFullAccessPath();
+  // TODO: Probably buggy--thinks "import Foo" == "import Foo.Bar".
+  // ImportPathBase should provide universal comparison functions to avoid this.
```
If you concur with this comment, I'll be happy to add an SR number. (I don't want to make functional changes in this PR if possible.)

nathawes replied:
Seems like a bug to me too.

brentdax replied:
https://bugs.swift.org/browse/SR-13490

I'll add the bug number to the comment so we don't lose track of it.

brentdax replied:
I've added the bug number to the code.

brentdax commented on lib/IDE/SourceEntityWalker.cpp:
```diff
@@ -644,14 +644,15 @@ bool SemaAnnotator::handleImports(ImportDecl *Import) {
-    return false;
+  if (Path.hasSubmodule())
+    if (!passModulePathElements(Path.takeParent(), ClangMod->Parent))
+      return false;
```
`ImportPath::Module` enforces an invariant that it has at least one module name in it, which required restructuring here. I think I've preserved the old logic, but it's hard to tell—I'm notsure when `ClangMod` can be nullptr or when `Path` could previously be empty.

nathawes replied:
This looks correct to me too. I think the empty `Path` handling was just related to the recursive call here, since it didn't have an equivalent of the `hasSubmodule` check before and would `drop_back` the last path component and get the non-existent parent of the corresponding `ClangMod`.

brentdax replied:
I've tweaked this a bit to flatten the control flow and enforce an invariant that a null `ClangMod` should never be passed in.

nathawes replied:
Thank you!

brentdax commented on lib/IRGen/IRGenDebugInfo.cpp:
```diff
@@ -2088,7 +2088,11 @@ void IRGenDebugInfoImpl::emitImport(ImportDecl *D) {
+  // ModuleDecl::ImportedModule inconsistently with typical use (placing the
+  // module path in the `accessPath` field). It's not immediately clear if there
+  // are matching inconsistencies elsewhere in IRGenDebugInfo.
+  ModuleDecl::ImportedModule Imported = { ImportPath::Access(), D->getModule() };
```
I don't think anything here reads the `accessPath` field (so why are we passing `ImportedModule`s around instead of just `ModuleDecl*`s?), but I could use an extra eye on this.

brentdax replied:
@adrian-prantl Can you comment on this?

adrian-prantl replied:
This is used only in IRGenDebugInfoImpl::finalize() to avoid creating a debug info breadcrumb for a module import twice. If you are saying this use is inconsistent, I'd image that it likely doesn't work correctly at the moment. We probably didn't notice, since a duplicate import in the debug info won't cause any problems other than wasting space.

Can you remind me what the access path is in a typical module import example? This was never really clear to me since Swift uses a flat namespace for modules. Or is that changing now?

brentdax replied:
In a scoped import (like `import class Foo.Bar`), the access path is the declaration the import is scoped to (`Bar`). Most imports are not scoped (like `import Foo`); their access paths are empty.

If you’re trying to record what was imported so that you can fully(ish) recreate it later, you probably want to write `D->getAccessPath()` instead of `ImportPath::Access()` here. If you just want to record the modules and don’t care about the access paths, what’s here is fine.

adrian-prantl replied:
Okay, that makes sense! At the moment, we only record imports on a module granularity in the debug info <rdar://problem/68149326>, so I think having an empty access path is actually what we'd want here!

brentdax replied:
Since you might eventually decide to start recording access paths, I think I'll just grab the right value from the ImportDecl and pass it along. Currently you'll just ignore it, but if that changes in the future you won't need to update this code.

brentdax replied:
I've force-pushed a change to construct it with the correct (ignored) access path.

brentdax commented on lib/Serialization/ModuleFile.cpp:
```diff
@@ -219,17 +212,7 @@ Status ModuleFile::associateWithFileContext(FileUnit *file, SourceLoc diagLoc) {
-          ctx.AllocateCopy(llvm::makeArrayRef(accessPathElem)), module};
-    }
+    dependency.Import = ModuleDecl::ImportedModule{accessPath, module};
 
```
The section between line 190/190 and line 232/215 has been significantly restructured and deserves special attention.

brentdax replied:
@xymus I'd appreciate if you could review at least this portion of the PR (plus the other one I'll ping you in).

xymus replied:
These changes look good and it's a nice improvement to readability.

brentdax commented on lib/Serialization/ModuleFile.cpp:
```diff
@@ -520,12 +495,10 @@ void ModuleFile::getImportDecls(SmallVectorImpl<Decl *> &Results) {
 
       auto *ID = ImportDecl::create(Ctx, FileContext, SourceLoc(), Kind,
-                                    SourceLoc(), AccessPath);
+                                    SourceLoc(), importPath.get());
```
The section between line 481/464 and line 528/501 has been significantly restructured and deserves special attention.

brentdax replied:
@xymus I'd appreciate if you could review at least this portion of the PR (plus the other one I'll ping you in).

xymus replied:
Same here. 👍

brentdax commented on utils/gyb_syntax_support/DeclNodes.py:
```diff
@@ -480,12 +480,14 @@
                    is_optional=True),
          ]),
 
+    # FIXME: technically misnamed; should be "ImportPathComponent"
     Node('AccessPathComponent', kind='Syntax',
          children=[
             Child('Name', kind='IdentifierToken'),
             Child('TrailingDot', kind='PeriodToken', is_optional=True),
          ]),
 
+    # FIXME: technically misnamed; should be "ImportPath"
     Node('AccessPath', kind='SyntaxCollection', element='AccessPathComponent'),
```
Fixing this will change names in SwiftSyntax; I'm not sure that's quite worth it.

brentdax commented:
@nathawes I'd appreciate it if you could review the changes to lib/IDE, and particularly the bits where I've left comments.

brentdax commented:
I didn't see any breakage in downstream projects, but I don't quite believe that, so I'll start with a smoke test to confirm it.

@swift-ci please smoke test

brentdax commented:
```
17:23:01 /home/buildnode/jenkins/workspace/swift-PR-Linux-smoke-test/branch-master/llvm-project/lldb/source/Plugins/Language/Swift/SwiftLanguage.cpp:1095:36: error: no type named 'AccessPathTy' in 'swift::ModuleDecl'
17:23:01                 swift::ModuleDecl::AccessPathTy access_path;
17:23:01                 ~~~~~~~~~~~~~~~~~~~^
17:23:01 1 error generated.
```
I knew it was too good to be true.

hamishknight commented on include/swift/AST/Import.h:
```diff
@@ -35,6 +43,282 @@ enum class ImportKind : uint8_t {
+                );
+    }
+
+    Subclass takeTopLevel() const {
+      assert(size() >= 1 && "nothing to take");
+      return Subclass(raw.take_front());
+    }
+
+    Subclass takeParent() const {
+      assert(size() >= 0 && "nothing to take");
+      return Subclass(raw.drop_back());
+    }
```
Nit: "take" usually indicates a consuming operation, perhaps these should be `getTopLevel` and `getParent`?

brentdax replied:
Hmm, maybe. I was basing this on the underlying `take_front()`. 

(At one point, I had `topLevel()` instead of `front()`, but I changed my mind about this because it caused too much friction. `takeTopLevel()` was to contrast with `topLevel()`.

brentdax replied:
I've changed these to `getTopLevelPath()` and `getParentPath()`.

hamishknight commented on include/swift/AST/Import.h:
```diff
@@ -0,0 +1,350 @@
+  /// import.
+  Access getAccessPath(bool isScoped) const {
+    if (isScoped)
+      return Access(getRaw().take_back());
```
Should this assert that there's at least two elements in the path? Just to make sure we don't accidentally pick up a module name

brentdax replied:
Good idea.

hamishknight reviewed (approved):
Neat!

brentdax commented on lib/ClangImporter/ClangImporter.cpp:
```diff
@@ -1783,15 +1783,15 @@ ModuleDecl *ClangImporter::Implementation::loadModuleClang(
   // Verify that the submodule exists.
   clang::Module *submodule = clangModule;
-  for (auto &component : path.slice(1)) {
+  for (auto &component : path.getRaw().slice(1)) {
```
Yes, it could. Good catch.

nathawes reviewed (approved):
The SourceKit/IDE bits look good to me. Thanks!

brentdax commented:
@swift-ci test

brentdax commented:
@swift-ci please smoke test adn merge

brentdax commented:
@swift-ci please smoke test and merge

brentdax commented:
@swift-ci please smoke test macOS platform

brentdax commented:
@swift-ci please smoke test macOS platform

beccadax committed:
[NFC] Adopt new ImportPath types and terminology

# Conflicts:
#	lib/IDE/CodeCompletion.cpp

brentdax commented:
@swift-ci please smoke test and merge

brentdax commented:
@swift-ci please smoke test and merge
